auth/authentication.go

```diff
@@ -17,119 +16,142 @@ type Database interface {
 	GetApplicationByToken(token string) (*model.Application, error)
 	GetClientByToken(token string) (*model.Client, error)
 	GetPluginConfByToken(token string) (*model.PluginConf, error)
+
 	GetUserByName(name string) (*model.User, error)
 	GetUserByID(id uint) (*model.User, error)
+	CreateUser(user *model.User) error
+}
+
+// AuthenticationProvider provides authentication methods
+type AuthenticationProvider interface {
+	Name() string
+	Authenticate(req *http.Request) (user *model.User, err error)
 }
 
 // Auth is the provider for authentication middleware
 type Auth struct {
-	DB Database
+	authenticators []AuthenticationProvider
+	DB             Database
 }
 
-type authenticate func(tokenID string, user *model.User) (authenticated bool, success bool, userId uint, err error)
+// RegisterAuthenticationProvider registers a new authentication provider
+// the provider is used in the order it i registered
+func (a *Auth) RegisterAuthenticationProvider(auth AuthenticationProvider) {
+	a.authenticators = append(a.authenticators, auth)
+}
 
-// RequireAdmin returns a gin middleware which requires a client token or basic authentication header to be supplied
+func (a *Auth) getUserInfoFromAuth(ctx *gin.Context) (user *model.User, authenticator string, err error) {
+	designatedAuthenticator := ctx.GetHeader("X-Gotify-Authenticator")
```

Comment:
Is it now required to specify this header? This creates a security risk because a user that exists in our internal auth, could be authenticated via a different origin.

I thought we go through the authenticators one by one and try to authenticate the user.
* if wrong password -> abort request
* if unknown user -> goto next authenticator

Reply:
I think we are on the same track.

> if wrong password -> abort request
Authenticators can return an error, this would abort the authentication process.

> if unknown user -> goto next authenticator
On line 48, if `designatedAuthenticator == ""`, all authenticator would be tried sequentially, otherwise, only authenticators with matching name will be used.

Reply:
Ahh, I missed the condition where user and err is nil, can we make this more explicit? Could be by returning a named error like ErrUserNotFound = errors.new("user not found"), or by adding another else branch
```
		} else if err != nil {
			return nil, providerName, err
		} else {
			// continue with next provider
		}
```

Reply:
I think we can't let external plugins import our auth package as it would introduce an import loop. The way around, as I see, would be to introduce another method to get whether this is a fatal authentication error in our error interface like maybe `Abort() bool` and default this to `false` if this method is missing.

Reply:
The error constant could be inside the plugin api. But I'm okay with both ways.

Reply:
I think an error constant inside the plugin API is the better way because it's easier to extend the API afterwards without breaking existing code.

Reply:
I don't think it is necessary to introduce another common api package just for authenticators as we did in user plugins, as we don't have as many data types to take care of.